File: modules/core/src/main/java/kr/entree/enderwand/collection/Reader.java
package kr.entree.enderwand.collection;


import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Created by JunHyung Lim on 2020-02-04
 */
public class Reader<T> {
    private final List<T> list;
    private int pos = 0;

    public static <T> Reader<T> of(List<T> list) {
        return new Reader<>(list);
    }

    @SafeVarargs
    public static <T> Reader<T> of(T... values) {
        return of(Lists.of(values));
    }

    private Reader(List<T> list) {
        this.list = list;
    }

    public int position() {
        return pos;
    }

    public boolean canRead() {
        return list.size() > pos;
    }

    public T peek() {
        return list.get(pos);
    }

    @Nullable
    public T peekOrNull() {
        return canRead() ? peek() : null;
    }


}

File: modules/core/src/main/kotlin/kr/entree/enderwand/collection/Reader.kt
package kr.entree.enderwand.collection

/**
 * Created by JunHyung Lim on 2019-12-03
 */
fun <T> List<T>.toReader() = Reader(this)

fun <T> Array<T>.toReader() = toList().toReader()

class Reader<T>(
    private val list: List<T>
) : List<T> by list {
    var pos = 0

    fun canRead() = size > pos

    fun peek() = list[pos]

    fun peekOrNull() = if (canRead()) peek() else null

    fun read() = list[pos++]

    fun readOrNull() = if (canRead()) read() else null

    fun remain() = (size - pos).coerceAtLeast(0)
}